internal/v4/api.go

```diff
@@ -850,14 +871,30 @@ func (h *Handler) serveAdminPromulgate(id *charm.Reference, _ bool, w http.Respo
 	return nil
 }
 
-func (h *Handler) authIdHandler(f router.IdHandler) router.IdHandler {
-	return func(charmId *charm.Reference, fullySpecified bool, w http.ResponseWriter, req *http.Request) error {
-		if err := h.authorizeEntity(charmId, req); err != nil {
+type resolvedIdHandler func(id *router.ResolvedURL, fullySpecified bool, w http.ResponseWriter, req *http.Request) error
+
+func (h *Handler) authId(f resolvedIdHandler) resolvedIdHandler {
+	return func(id *router.ResolvedURL, fullySpecified bool, w http.ResponseWriter, req *http.Request) error {
+		if err := h.authorizeEntity(id, req); err != nil {
 			return errgo.Mask(err, errgo.Any)
 		}
-		if err := f(charmId, fullySpecified, w, req); err != nil {
+		if err := f(id, fullySpecified, w, req); err != nil {
 			return errgo.Mask(err, errgo.Any)
 		}
 		return nil
 	}
 }
+
+func isFullySpecified(id *charm.Reference) bool {
+	return id.Series != "" && id.Revision != -1
+}
+
+func (h *Handler) resolveId(f resolvedIdHandler) router.IdHandler {
+	return func(id *charm.Reference, w http.ResponseWriter, req *http.Request) error {
+		rid, err := h.resolveURL(id)
```

Comment:
I like that the id handlers can now decide if and when to resolve the URL themselves.
This approach is much more flexible, thanks a lot!